core/rawdb/freezer.go

```diff
@@ -255,35 +258,61 @@ func (f *freezer) freeze(db ethdb.KeyValueStore) {
 	nfdb := &nofreezedb{KeyValueStore: db}
 
 	for {
+		select {
+		case <-f.quit:
+			log.Info("Freezer shutting down")
+			return
+		default:
+		}
 		// Retrieve the freezing threshold.
 		hash := ReadHeadBlockHash(nfdb)
 		if hash == (common.Hash{}) {
 			log.Debug("Current full block hash unavailable") // new chain, empty database
-			time.Sleep(freezerRecheckInterval)
-			continue
+			select {
+			case <-time.NewTimer(freezerRecheckInterval).C:
+				continue
+			case <-f.quit:
+				return
+			}
```

Comment:
Doing these sleeps inlined like this looks a bit .. not so elegant :) 
Perhaps you can set a `backoff`-flag, and do the sleep/select on just one place?

Reply:
Good that you added these btw -- I was going to write further above that "hey the close might now take a minute", but then I scrolled down and saw these :)

Reply:
I figured out that it might take a minute the hard way...

Sure, I'll switch it over to use a backoff flag.